src/experiment.ts

```diff
@@ -0,0 +1,127 @@
+import { Identify } from 'amplitude-js';
+import { Client } from './client';
+
+const LOCAL_STORAGE_EXPERIMENTS_PREFIX = '__analytics_experiments_';
+
+/** Experiment interface allows getting a variation for a specified user. */
+export interface Experiment<Variation extends string> {
+	name: string;
+
+	engage(userId: string): Variation;
+}
+
+interface VariationData<Variation> {
+	variation: Variation;
+	fraction: number;
+	margin: number;
+}
+
+/** LocalExperiment allows describing an Experiment implemented with local storage. */
+export class LocalExperiment<Variation extends string>
+	implements Experiment<Variation> {
+	private readonly data: Array<VariationData<Variation>> = [];
+	private defaultVariation?: Variation;
+
+	constructor(
+		public readonly name: string,
+		private readonly analytics?: Client,
+	) {}
+
+	private static checkPercent(f: number) {
+		if (isNaN(f) || f < 0 || f > 100) {
+			throw new Error(
+				'Variation fraction must be defines as a percent value between 0 and 100',
+			);
+		}
+	}
+
+	private checkDuplicates(variation: Variation) {
+		const present = this.data.find(d => d.variation === variation);
+		if (present != null) {
+			throw new Error(
+				`Variation [${present.variation} ${present.fraction}%] already exists in experiment ${this.name}.`,
+			);
+		}
+	}
+
+	private static dataString(data: Array<VariationData<any>>): string {
+		return data.map(d => `variation ${d.variation}: ${d.fraction}%`).join(', ');
+	}
+
+	define(variation: Variation, fraction: number): LocalExperiment<Variation> {
+		LocalExperiment.checkPercent(fraction);
+		this.checkDuplicates(variation);
+
+		let previousMargin = 0;
+		if (this.data.length > 0) {
+			previousMargin = this.data[this.data.length - 1].margin;
+		}
+		const margin = previousMargin + fraction;
+
+		const varData = { variation, margin, fraction };
+		if (margin > 100) {
+			const allData = LocalExperiment.dataString(this.data.concat(varData));
+			throw new Error(
+				`Incorrect fraction in experiment ${this.name}. Sum of fractions is greater than 100%: ${allData}`,
+			);
+		}
+
+		this.data.push(varData);
+		return this;
+	}
+
+	defineDefault(variation: Variation): LocalExperiment<Variation> {
+		this.defaultVariation = variation;
+		return this;
+	}
+
+	engage(userId: string): Variation {
+		if (this.data.length === 0) {
+			throw new Error(`Variations are not defined for experiment ${this.name}`);
+		}
+		if (
+			this.data[this.data.length - 1].margin < 100 &&
+			this.defaultVariation == null
+		) {
+			throw new Error(
+				`Experiments is not fully defined. Current data: ${LocalExperiment.dataString(
+					this.data,
+				)}`,
+			);
+		}
+
+		if (window.localStorage == null) {
```

Comment:
On non-browser environments `window` will be undefined, so 
```suggestion
		if (window?.localStorage == null) {
```